subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/VersionRangeResolveIntegrationTest.groovy

```diff
@@ -0,0 +1,539 @@
+        "[1.0, 1.2]" | "[1.0, )"    | "FAIL"        | "FAIL"       | ''             | ''
+        "[1.0, )"    | "1.0"        | "1.2"         | "1.2"        | ''             | ''
+        "[1.0, )"    | "[1.0, 1.1]" | "1.2"         | "1.2"        | ''             | ''
+        "[1.0, )"    | "[1.0, 1.4]" | "FAIL"        | "FAIL"       | ''             | ''
+        "[1.0, )"    | "[1.0, )"    | "FAIL"        | "FAIL"       | ''             | ''
+        "1.+"        | "1.0"        | "1.2"         | "1.2"        | ''             | ''
+        "1.+"        | "[1.0, 1.1]" | "1.2"         | "1.2"        | ''             | ''
+        "1.+"        | "[1.0, 1.4]" | "FAIL"        | "FAIL"       | ''             | ''
+        "1.+"        | "[1.0, )"    | "FAIL"        | "FAIL"       | ''             | ''
+
+        // Incorrect behaviour: should find older version in preferred range when newest is rejected
+        "[1.0, 1.2]" | "1.2"        | "FAIL"        | "FAIL"       | "1.1"          | "1.1"
+        "[1.0, )"    | "1.2"        | "FAIL"        | "FAIL"       | "1.1"          | "1.1"
+        "1.+"        | "1.2"        | "FAIL"        | "FAIL"       | "1.1"          | "1.1"
+    }
+
+
+    @Unroll
+    def "resolve #one & #two"() {
+        expect:
+        resolve(one, two) == lenientResult
+        resolve(strict(one), two) == strict1Result
+        resolve(one, strict(two)) == strict2Result
+        resolve(strict(one), strict(two)) == strictBothResult
+
+        resolve(two, one) == lenientResult
+        resolve(two, strict(one)) == strict1Result
+        resolve(strict(two), one) == strict2Result
+        resolve(strict(two), strict(one)) == strictBothResult
+
+        where:
+        one         | two         | lenientResult | strict1Result | strict2Result | strictBothResult
+        FIXED_7     | FIXED_13    | 13            | -1            | 13            | -1
+        FIXED_12    | FIXED_13    | 13            | -1            | 13            | -1
+        FIXED_12    | RANGE_10_11 | 12            | 12            | -1            | -1
+        FIXED_12    | RANGE_10_14 | 12            | 12            | 12            | 12
+        FIXED_12    | RANGE_13_14 | 13            | -1            | 13            | -1
+        FIXED_12    | RANGE_7_8   | -1            | -1            | -1            | -1
+        FIXED_12    | RANGE_14_16 | -1            | -1            | -1            | -1
+        RANGE_10_11 | FIXED_10    | 10            | 10            | 10            | 10
+        RANGE_10_14 | FIXED_13    | 13            | 13            | 13            | 13
+        RANGE_10_14 | RANGE_10_11 | 11            | 11            | 11            | 11
+        RANGE_10_14 | RANGE_10_16 | 13            | 13            | 13            | 13
+    }
+
+    @Unroll
+    def "resolve #one & #two & #three"() {
+        expect:
+        [one, two, three].permutations().each {
+            assert resolve(it[0], it[1], it[2]) == rst
```

Comment:
Actually I'd suggest replacing this whole test with something like this, which creates a Spock iteration per thing to test. The reason is that the current implementation will fail _as soon as one permutation fails_, making it hard to understand what are the actual failing test cases. Instead, the custom generator creates a Spock iteration per permutation, and it makes it very clear in which cases we fail:

```groovy
    @Unroll
    def "resolve #permutation"() {
        given:
        def candidates = permutation.candidates
        def expected = permutation.expected

        expect:
        resolve(candidates) == expected

        where:
        permutation << StrictPermutationsProvider.check(
            versions: [FIXED_12, FIXED_13, FIXED_10],
            expectedNoStrict: 13,
            expectedStrict: [-1, 13, -1]
        ).and(
            ignore: true,
            versions: [FIXED_10, FIXED_12, RANGE_10_14],
            expectedNoStrict: 12,
            expectedStrict: [-1, 12, 12]
        ).and(
            ignore: true,
            versions: [FIXED_10, RANGE_10_11, RANGE_10_14],
            expectedNoStrict: 10,
            expectedStrict: [10, 10, 10]
        ).and(
            ignore: true,
            versions: [FIXED_10, RANGE_11_12, RANGE_10_14],
            expectedNoStrict: 12,
            expectedStrict: [-1, 12, 12]
        ).and(
            versions: [FIXED_10, RANGE_10_11, RANGE_13_14],
            expectedNoStrict: 13,
            expectedStrict: [-1, -1, 13]
        ).and(
            ignore: true,
            versions: [RANGE_10_11, RANGE_10_12, RANGE_10_14],
            expectedNoStrict: 11,
            expectedStrict: [11, 11, 11]
        ).and(
            versions: [RANGE_10_11, RANGE_10_12, RANGE_13_14],
            expectedNoStrict: 13,
            expectedStrict: [-1, -1, 13]
        ).and(
            versions: [RANGE_10_11, RANGE_10_12, RANGE_13_14],
            expectedNoStrict: 13,
            expectedStrict: [-1, -1, 13]
        ).and(
            ignore: true,
            description: "gradle/gradle#4608",
            versions: [FIXED_10, FIXED_10, FIXED_12],
            expectedNoStrict: 12,
            expectedStrict: [-1, -1, 12]
        ).and(
            ignore: true,
            description: "gradle/gradle#4608",
            versions: [FIXED_12, RANGE_12_14, RANGE_10_11],
            expectedNoStrict: 12,
            expectedStrict: [12, 12, -1]
        ).and(
            ignore: true,
            description: "gradle/gradle#4608",
            versions: [FIXED_12, RANGE_12_14, FIXED_10],
            expectedNoStrict: 12,
            expectedStrict: [12, 12, -1]
        )
    }

    static class StrictPermutationsProvider implements Iterable<Candidate> {
        private final List<Batch> batches = []
        private int batchCount

        static StrictPermutationsProvider check(Map config) {
            new StrictPermutationsProvider().and(config)
        }

        StrictPermutationsProvider and(Map config) {
            assert config.versions
            assert config.expectedNoStrict
            assert config.expectedStrict
            if (!config.ignore) {
                List<RenderableVersion> versions = config.versions
                List<Integer> expectedStrict = config.expectedStrict
                List<Batch> iterations = []
                String batchName = config.description ?: "#${batchCount++} : ${versions}"
                iterations.add(new Batch(batchName, versions, config.expectedNoStrict))
                versions.size().times { idx ->
                    iterations.add(new Batch(batchName, versions.withIndex().collect { RenderableVersion version, idx2 ->
                        if (idx == idx2) {
                            def v = new StrictVersion()
                            v.version = version.version
                            v
                        } else {
                            version
                        }
                    }, expectedStrict[idx]))
                }
                batches.addAll(iterations)
            }

            this
        }

        @Override
        Iterator<Candidate> iterator() {
            new PermutationIterator()
        }

        @Canonical
        static class Batch {
            String batchName
            List<RenderableVersion> versions
            int expected
        }

        class PermutationIterator implements Iterator<Candidate> {
            Iterator<Batch> batchesIterator = batches.iterator()
            String currentBatch
            Iterator<List<RenderableVersion>> current
            int expected

            @Override
            boolean hasNext() {
                batchesIterator.hasNext() || current?.hasNext()
            }

            @Override
            Candidate next() {
                if (current?.hasNext()) {
                    return new Candidate(batch: currentBatch, candidates:current.next() as RenderableVersion[], expected:expected)
                }
                Batch nextBatch = batchesIterator.next()
                expected = nextBatch.expected
                current = nextBatch.versions.permutations().iterator()
                currentBatch = nextBatch.batchName
                return next()
            }
        }

        static class Candidate {
            String batch
            RenderableVersion[] candidates
            int expected

            @Override
            String toString() {
               "Batch $batch " + candidates.collect { it.toString() }.join(' & ') + " -> ${expected<0?'FAIL':expected}"
            }
        }
    }
```

Reply:
Thanks for the ideas. This test is very much a work in progress, and was only included in this PR because it got tangled up with the changes (and I didn't bother detangling). I'll make your suggested improvements in a later PR.